bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,913 @@
+     * <p>
+     * The thread listens to the TCP socket opened at instantiation of the {@link QbusCommunication} class
+     * and interprets all incomming json messages. It triggers state updates for active channels linked to the
+     * Qbus outputs. It is started after initialization of the communication.
+     *
+     * @return
+     *
+     */
+    private void qEvents() {
+        String qMessage;
+
+        logger.info("Listening for events on thread {}", Thread.currentThread().getId());
+        listenerStopped = false;
+        qEventsRunning = true;
+
+        BufferedReader reader = qIn;
+
+        try {
+            if (reader == null) {
+                throw new IOException("Bufferreader for incomming messages not initialized.");
+            }
+            while (!Thread.currentThread().isInterrupted() & ((qMessage = reader.readLine()) != null)) {
+                if (qMessage != null) {
+                    readMessage(qMessage);
+                }
+            }
+        } catch (IOException e) {
+            if (!listenerStopped) {
+                qEventsRunning = false;
+                logger.error("Qbus: IO error in listener on thread {}", Thread.currentThread().getId());
+
+                QbusBridgeHandler handler = bridgeCallBack;
+
+                if (handler != null) {
+                    CTDConnected = false;
+                    handler.bridgeOffline("No communication with Qbus server");
+                }
+            }
+        }
+
+        qEventsRunning = false;
+        logger.trace("Event listener thread stopped on thread {}", Thread.currentThread().getId());
+    };
+
+    /**
+     * Called by other methods to send json data to Qbus.
+     *
+     * @param qMessage
+     * @throws InterruptedException
+     */
+    synchronized void sendMessage(Object qMessage) throws InterruptedException {
+        PrintWriter writer = qOut;
+        String json = gsonOut.toJson(qMessage);
+
+        if (writer != null) {
+            writer.println(json);
+            // Delay after sending data to improve scene execution
+            try {
+                TimeUnit.MILLISECONDS.sleep(250);
+            } catch (InterruptedException e) {
+                throw new InterruptedException(e.toString());
+            }
+
+        }
+        if ((writer == null) || (writer.checkError())) {
+            logger.warn("Error sending message, trying to restart communication");
+            restartCommunication();
+            // retry sending after restart
+            writer = qOut;
+            if (writer != null) {
+                writer.println(json);
+            }
+            if ((writer == null) || (writer.checkError())) {
+                logger.warn("Error resending message");
+
+            }
+        }
+    }
+
+    /**
+     * Called by other methods to Qbus server and read response
+     */
+    private void sendAndReadMessage(String command) throws IOException, InterruptedException {
+        @Nullable
```

Comment:
The compiler is more intelligent on local variables. Therefore, the Nullable could be removed.